src/Lecture2.hs

```diff
@@ -268,7 +366,21 @@ data EvalError
 It returns either a successful evaluation result or an error.
 -}
 eval :: Variables -> Expr -> Either EvalError Int
-eval = error "TODO"
+eval         _ (Lit lit) = Right lit
+eval variables (Var var) = case (lookup var variables) of
+                             Just int -> Right int
+                             Nothing  -> Left (VariableNotFound var)
+eval variables (Add expr1 expr2)
+  | isRight result1 && isRight result2 = Right $ (fromRight' result1 + fromRight' result2)
+  | isLeft result1                     = result1
+  | isLeft result2                     = result2
+  | otherwise                          = error "Computational error"
+    where result1 = eval variables expr1
+          result2 = eval variables expr2
+
+fromRight' :: Either a b -> b
+fromRight' (Left _) = error "Computational error"
+fromRight' (Right b) = b
```

Comment:
That's a possible impementation.

But even without monads, you can use pattern matching to write this function more simpler. You're following the "Boolean blindness" antipattern where you first check the constructor and then extract with a partial functions.

Instead, you could use pattern-matching on the values of type`Either`:

```haskell
eval variables (Add expr1 expr2) = append (eval variables expr1) (eval variables expr2)
  where
    append :: Either EvalError Int -> Either EvalError Int
    append (Left e) _ = Left e
    append _ (Left e) = Left e
    spend (Right x) (Right y) = Right (x + y)
```

And nce you learn about Monads, you'll be able to reduce this entire computation to a single line ✂️